zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
-                            end);
-
-        return instance;
+    zcm::Ptr{Native.Zcm}
+    provider::String
+    subscriptions::Vector{Subscription}
+
+    function Zcm(provider::AbstractString = "")
+        pointer = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), provider);
+        instance = new(pointer, provider, Subscription[])
+        finalizer(instance, close)
+        return instance
     end
+end
 
-    function Zcm() return Zcm(""); end
+function close(zcm::Zcm)
+    if zcm.zcm != C_NULL
+        ccall(("zcm_destroy", "libzcm"), Void,
+              (Ptr{Native.Zcm},), zcm)
+        zcm.zcm = C_NULL
+    end
 end
-export Zcm;
 
-function encode end
-function decode end
+# Defines the conversion when we pass this to a C function expecting a pointer
+Base.unsafe_convert(::Type{Ptr{Native.Zcm}}, zcm::Zcm) = zcm.zcm
 
-function good(zcm::Zcm)
-    return (zcm.zcm != C_NULL) && (errno(zcm) == 0);
-end
-export good;
+good(zcm::Zcm) = (zcm.zcm != C_NULL) && (errno(zcm) == 0)
 
-function errno(zcm::Zcm)
-    return ccall(("zcm_errno", "libzcm"), Cint, (Ptr{Native.Zcm},), zcm.zcm);
-end
-export errno;
+errno(zcm::Zcm) = ccall(("zcm_errno", "libzcm"), Cint, (Ptr{Native.Zcm},), zcm)
 
 function strerror(zcm::Zcm)
-    val =  ccall(("zcm_strerror", "libzcm"), Cstring, (Ptr{Native.Zcm},), zcm.zcm);
+    val =  ccall(("zcm_strerror", "libzcm"), Cstring, (Ptr{Native.Zcm},), zcm)
     if (val == C_NULL)
-        return "unable to get strerror";
+        return "unable to get strerror"
     else
-        return unsafe_string(val);
+        return unsafe_string(val)
     end
 end
-export strerror;
 
-# TODO: get this into docs:
-# Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, handler, usr)
-
-    handler_jl = __zcm_handler(handler, usr);
-    handler_c  = cfunction(__zcm_handler_wrapper, Void,
-                           (Ptr{Native.RecvBuf}, Cstring, Ptr{Void},));
-    uv_wrapper = ccall(("uv_zcm_msg_handler_create", "libzcmjulia"), Ptr{Native.UvSub},
-                       (Ptr{Void}, Any,), handler_c, handler_jl);
-    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"));
-
-    return Sub(ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
-                     (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub},),
-                     zcm.zcm, channel, uv_handler, uv_wrapper),
-               handler_jl, handler_c, uv_handler, uv_wrapper);
+function onresponse(rbuf::Native.RecvBuf, channelbytes::Cstring, opts::SubscriptionOptions)
+    channel = unsafe_string(channelbytes)
+    msgdata = unsafe_wrap(Vector{UInt8}, rbuf.data, rbuf.data_size)
+    msg = decode(opts.msgtype, msgdata)
+    opts.handler(channel, msg)
+    return nothing
+end
+
+function subscribe(zcm::Zcm, channel::AbstractString, options::T) where T <: SubscriptionOptions
```

Comment:
How do the `handler`s in the example files get converted to `options` here?

Reply:
Ahh i spoke too soon. I see the other subscribe function below. Why have 2? And why not just have `options` be of type `SubscriptionOptions`? Why inherit?

Reply:
Actually I might just not understand this `where` syntax at all haha. Why not just do `options::SubscriptionOptions`?

Reply:
The `<:` is used here to denote that this `subscribe` method only applies when the `options` argument has any concrete type `T` that is a subtype of `SubscriptionOptions` (recall that `SubscriptionOptions` itself is not a concrete type, as it has free type parameters). The type `T` is needed later on in the function, so this is a convenient way to get it. The above is (apart from a technicality that isn't really relevant here) equivalent to:
```julia
function subscribe(zcm::Zcm, channel::AbstractString, options::SubscriptionOptions)
    T = typeof(options)
    # rest of function
end
```

Reply:
Ahh interesting. So much syntax to learn, so little time

Reply:
Is this new in 0.6? Seems to not be working for me

Reply:
ah, sorry, yes, this is only in v0.6 and higher. I'll switch this to v0.5-compatible syntax so you can try it (although I still suggest moving to v0.6 anyway)

Reply:
resolved in https://github.com/ZeroCM/zcm/pull/165/commits/7c650529f78c7c972cab0849c1ccb546c5106e28